### ISSUE TITLE
Spellcasting and attacks should be separated to allow for filtering of spells

### ISSUE THREAD

**Issue by Pjb518** [UnicornHeartClub/Issues#119]
Spells have very different features to regular attacks, including casting times, ritual and concentration tags, components (verbal, material, somatic), school, level, etc. It would be useful to be able to filter spells using these special qualities to easily look up spells without a concentration tag, or to find rituals, etc. This kind of filter doesn't make sense if standard attacks are stored in the same place.

As such, I suggest we separate the attacks section from the spellcasting section, giving them their own tailored input fields, and filtering options.

**Comment by Aquaricat** [comment]
Rather than separating them, add a flag on each item that marks it as a melee or an attack (which could then display an indicator icon in the list, besides), and filter against that.
#AlternateIdeas

**Comment by Pjb518** [comment]
I'm not sure that's worth doing since the actual input fields we'd want for each section are also drastically different. I also wouldn't want to hide things like melee attacks just because a user is perusing their non concentration spell items. The spellcasting section also provides a nice place to display things like a spell attack bonus and the spell save DC which aren't present on the current sheets at all.

**Comment by Aquaricat** [comment]
Well, consider that weapons *can* have save DCs on them, for certain abilities, such as the Gulthias Staff from CoS.

Also, I'd argue that the Attacks/Spellcasting section really isn't the best place to show lots of extended information if it can be avoided.  When I look at the Core sheet, I usually find myself thinking of seeing at-a-glance type information, not highly detailed, searchable lists.

That said, I mean, there's room (to me) for some additional compressed information - each line could have a mouse-overable icon that shows a popup with this sort of additional information, for example

**Comment by Pjb518** [comment]
I'm not saying we should add a DC for each item, and things like DCs for weapons really don't have a place in my opinion. I'm talking about just the spell save DC and spell attack bonus for each class, like you might find at the top of the 5E spell sheets. As we only support one class characters right now, this is just going to be a single item, but in the future we could separate the DCs and attack bonuses for each class the character has, so they can give this information easily.

The shaped sheet has something along these lines, along with button toggles to filter the spell list. While I personally think this section is a total mess, I do think they're on the right track, and I know that a lot of players use this to quickly look over their spell list. Is x a concentration spell? Easy to find out.

![image](https://user-images.githubusercontent.com/33249712/34912356-d8acf130-f8df-11e7-9203-b6583e99b822.png)

The shaped sheet also keeps track of known and prepared spells which is handy so that you don't have to keep track yourself, which can be a little cumbersome for high level characters.

**Comment by Pjb518** [comment]
I actually like the idea of being able to see a breakdown the abilities on mouse-over, as you suggest, but the reason I'm asking for toggles is really completely different and not solved by that feature. Say you have 25 spells and you want to know what you can cast while still keeping up bless. Toggle only non-concentration spells. This gives you an at a glance breakdown of your options for the next turn, instead of having to mouse over each item, or even scan a list where there's a little C next to each item. It's much easier to just have the non-concentration items removed.

The same goes for things like bonus action spells. You've cast a cantrip and now you want to know what else you can do. Toggle only bonus action spells.

Maybe you have prepared and unprepared spells for your class. You want to see only what you have prepared, while keeping the other stuff on your sheet.

There are plenty of reasons you'd want to toggle to make the list easier to digest.

**Comment by Aquaricat** [comment]
See, I think that's my exact point, though.  What you're talking about (to me) *does* deserve its own section, but that section isn't what's addressed by the Attacks & Spells section *as it stands*.  The section currently on the sheet is IMHO primarily there to drive attack and spell macros.  An actual spell list, though, especially one supporting multiple classes and prepared/unprepared entries deserves its own tab, at the same level as the current Core and Bio tabs.  

![image](https://user-images.githubusercontent.com/7004392/34912453-89912cca-f8a7-11e7-9d1e-0b96cb33eb83.png)

Splitting the Attacks/Spells section would actually make it a little more confusing for less savvy users if the proposed template activation ends up being adopted.  Now instead of just `!template name @me.attacks.id`, you're asking them to differentiate between the two - `!template name @me.spells.id` - for what is essentially the same information.  There's so much overlap between attacks and spells when it comes to defining them that IMHO it doesn't make sense programmatically to split them.

Of course, there's the obvious downside of my argument too - how do you then avoid duplicating information from the combined attacks/spells list, and the more detailed spell list tab.  Truth is, I'm not entirely sure.  I'm not entirely a fan of the R20 OGL sheet's solution, where adding a spell on the spell list tab creates a linked entry on the core tab, because there's a lot of spells where I won't ever need to click on that macro - it just becomes extra cruft on the sheet.  Fortunately you can delete the linked duplicate, but if I'm being honest I do 90% of my sheet keeping in DiceCloud anyway, so I never use R20's spell keeping features.

Also, there's a big unknown here still too: How are the Compendiums going to work in relationship to this kind of management?  /eye @tlackemann.  ^_-

**Comment by Pjb518** [comment]
I don't mind having a separate spells page at all, and it might actually be a better solution, but `!template name @me.spells.id` actually makes _a lot_ more sense to me, because referring to things like _aid_, _bless_, _longstrider_, etc. as attacks is pretty unintuitive, in my opinion.

**Comment by Aquaricat** [comment]
Meh, that's a low blow, bringing logic into this.  ^_-

That said, separating them into a second collection (which, yeah, does make more sense in the long run) could easily be done with that same flag I mentioned, and opens up the possibility of other collections if needed.  Abilities might make sense, for example.

**Comment by Pjb518** [comment]
For the attacks and spellcasting section, perhaps a little divider in the box could be a decent compromise, just so that it's easier to see where your abilities might be when using this as a launcher, essentially. Being able to split this up into spell level groups also would help here.

**Comment by Aquaricat** [comment]
Having that happen automatically would be awesome, and should be relatively easy.  Definitely gives it a bit more polish and makes it easier to navigate.

**Comment by Pjb518** [comment]
Yeah, absolutely. We're already entering spell levels for these abilities, so we can just place them based on that. Cantrip just counts as a spell level, and anything missing a spell level is just an attack.

We'd then just need to be able to set uses for things like innate casting options (drow magic, tiefling devilry, etc.) so that they don't consume slots automatically. Job done.
